### DISCUSSION TITLE
Show a single deep zoom image

### DISCUSSION THREAD

**Discussion by tomcrane** [digirati-co-uk/iiif-canvas-panel#4] [How Might We]
Continuing on from #1 - using the same pattern as #1, I should be able to specify that user can pan and zoom into the image, using typical/familiar controls/gestures.

Options 

Start as static image but switch to zoomable on interaction - click or zoom action
Show navigator overlay (as with OpenSeadragon)

**Comment by tomcrane** [comment]
More options

Make it easy for the page developer to prevent the component capturing scroll actions on the page and interpreting them as zoom actions within the viewport. If I have have one or more images on a page of editorial, and they fill a lot of the width (e.g., on a phone), it's annoying when my scroll actions don't scroll the page but get captured in an interactive image.

**Comment by stephenwf** [comment]
There are two parts here, the interaction design and the navigator. 

### Interaction design

The interaction is completely handled by a "controller" in Atlas. This is a standalone piece of code that edits 4 bytes of data (x, y, x2, y2 of the viewport on the image).

<details>
<summary>
  When you create a viewer you create a controller instance and pass it to the runtime.
</summary>


```js
const renderer = new Atlas.CanvasRenderer(canvas);
const viewport = { width: 800, height: 600, x: 0, y: 0, scale: 1 };
const world = new Atlas.World();

const controller = Atlas.popmotionController({
  minZoomFactor: 0.5,
  maxZoomFactor: 3,
  enableClickToZoom: false,
});

// Create our runtime.
const runtime = new Atlas.Runtime(renderer, world, viewport, [controller]);

// And start listening to browser events proxied from our div element.
new Atlas.BrowserEventManager(wrapper, runtime);
```
</details>

<details>
<summary>
In the code a controller is quite simple.
</summary>

```js
const myController = {
  start: function(runtime) {
    // your code..
    runtime.x = // ...
    runtime.y = // ...
    runtime.width = // ...
    runtime.height = // ...
  },
  stop: function(runtime) {},
};
```

It is passed a runtime object, which has many properties to read (or write). 

The most common way to add controller behaviour is using world events. These a _very_ similar to normal DOM events and should be compatible with libraries that use them.

```js
// The world has an API similar to the DOM. An additional `e.atlas` is added to each event with 
// a position in the atlas world. Everything else is the event is from the source (usually browser)
runtime.world.addEventListener('mousedown', e => {
   e.atlas.x; // x position in the world
   e.atlas.y // y position in the world
});
```

The purpose of a controller is to implement pan events, click to zoom, scroll-to-zoom and other interactions by translating events on the world to mutations of the viewport.
</details>

The default controller uses a library (popmotion) to handle inertia and velocity of panning and zooming. This is convenient as the math for calculating all of that is there and easy to use. However it does add to the bundle size. Ideally we would have a modular controller written in vanilla javascript/typescript that can handle smooth panning, zooming. 

### Renderer (for the navigator)

Another modular part of Atlas is the renderer. Currently there are a handful of renderers that will take the positions of images at a given frame and render them.
* **Static renderer** - Outputs HTML <img/> tags alongside HTML for non-images
* **Canvas renderer** - Draws onto an HTML5 canvas, supports CORS images
* **WebGL renderer** - Draws onto an HTML5 canvas with WebGL, does not support CORS images (but it's fast)
* **Overlay renderer** - Draws unsupported elements onto an HTML layer, used as a fallback for certain object types
* **Composite renderer** - Allows more than one renderer to be used at the same time.
* **Debug renderer** - Draws a mini representation of the items in the world, used for debugging (looks like a navigator without images)

A navigator would simply be another one of these. Instead of drawing only what's on screen it would draw the smallest representation of everything and an outline of the current viewport (similar to the debug).

**Reply by stephenwf** [reply]
Another potential renderer: SVGRenderer to produce very simple SVGs:
https://developer.mozilla.org/en-US/docs/Web/SVG/Element/image#example

```
<svg width="200" height="200"
  xmlns="http://www.w3.org/2000/svg">
  <image href="mdn_logo_only_color.png" height="200" width="200"/>
</svg>
```

Would also support quite a few things natively (CORS, pointer events, CSS styling etc.)

**Comment by stephenwf** [comment]
Atlas does not currently have native support for touch controls or pinch-zoom controls. This is something else that would be required for the basic case.

**Comment by stephenwf** [comment]
There is an experimental web components bundle for atlas:
```html
<atlas-viewer>
    <atlas-tiled-image src="https://iiif.bodleian.ox.ac.uk/iiif/image/5009dea1-d1ae-435d-a43d-453e3bad283f/info.json" />
</atlas-viewer>
```

**Reply by stephenwf** [reply]
Live demo: https://codesandbox.io/s/atlas-web-component-tile-bc2or?file=/index.html

**Reply by stephenwf** [reply]
And with static image: https://codesandbox.io/s/atlas-web-component-static-6z5tn?file=/index.html

**Comment by tomcrane** [comment]
Proposal

Extends the img-service and canvas-panel tags from #1 (comment)

This is _identical_ except that the render mode is zoom:

`<canvas-panel render="zoom" iiif-content=".." />`

This is, in fact, the default render mode, so is equivalent to:

`<canvas-panel iiif-content=".." />`

The programming interface is the same. 

In all render modes, you can constrain the viewport (the block space taken up by canvas-panel on the page) to a particular part of the canvas (or src image for img-service). This is covered in https://github.com/digirati-co-uk/iiif-canvas-panel/discussions/5. 

The difference in zoom mode is not so much that it switches to a tile source (it doesn't necessarily do this) but that the viewport is variable, usually under user control through panning and zooming actions. This exposes more events for the developer to handle.
